### ISSUE TITLE
Implement `pydoclint`

### ISSUE THREAD

**Issue by augustelalande** [astral-sh/ruff#12434]
This issue tracks the implementation of [pydoclint](https://jsh9.github.io/pydoclint/violation_codes.html). Following discussion in #11471 it was decided to prefer `pydoclint` over `darglint` for implementation is ruff since they cover mostly the same ruleset, and `pydoclint` is still actively developed (unlike `darglint`). This issue should then supersede #458.

## Rules
* [ ] `DOC101` Docstring contains fewer arguments than in function signature
* [ ] `DOC102` Docstring contains more arguments than in function signature
* [ ] `DOC103` Docstring arguments are different from function arguments
* [ ] `DOC104` Arguments are the same in the docstring and the function signature, but are in a different order
* [ ] `DOC105` Argument names match, but type hints do not match
* [ ] `DOC106` The option `--arg-type-hints-in-signature` is `True` but there are no argument type hints in the signature
* [ ] `DOC107` The option `--arg-type-hints-in-signature` is `True` but not all args in the signature have type hints
* [ ] `DOC108` The option `--arg-type-hints-in-signature` is `False` but there are argument type hints in the signature
* [ ] `DOC109` The option `--arg-type-hints-in-docstring` is `True` but there are no type hints in the docstring arg list
* [ ] `DOC110` The option `--arg-type-hints-in-docstring` is `True` but not all args in the docstring arg list have type hints
* [ ] `DOC111` The option `--arg-type-hints-in-docstring` is `False` but there are type hints in the docstring arg list
* [ ] `DOC201` Function/method does not have a return section in docstring
* [ ] `DOC202` Function/method has a return section in docstring, but there are no return statements or annotations
* [ ] `DOC203` Return type(s) in the docstring not consistent with the return annotation
* [ ] `DOC301` `__init__()` should not have a docstring; please combine it with the docstring of the class
* [ ] `DOC302` The class docstring does not need a "Returns" section, because `__init__()` cannot return anything
* [ ] `DOC303` The `__init__()` docstring does not need a "Returns" section, because it cannot return anything
* [ ] `DOC304` Class docstring has an argument/parameter section; please put it in the `__init__()` docstring
* [ ] `DOC305` Class docstring has a "Raises" section; please put it in the `__init__()` docstring
* [ ] `DOC306` The class docstring does not need a "Yields" section, because `__init__()` cannot yield anything
* [ ] `DOC307` The `__init__()` docstring does not need a "Yields" section, because `__init__()` cannot yield anything
* [ ] ~~`DOC401` (Deprecated; this violation code no longer appears)~~
* [ ] `DOC402` Function/method has "yield" statements, but the docstring does not have a "Yields" section
* [ ] `DOC403` Function/method has a "Yields" section in the docstring, but there are no "yield" statements, or the return annotation is not a Generator/Iterator/Iterable
* [ ] `DOC404` The types in the docstring's Yields section and the return annotation in the signature are not consistent
* [x] `DOC501` Function/method has "raise" statements, but the docstring does not have a "Raises" section
* [x] `DOC502` Function/method has a "Raises" section in the docstring, but there are not "raise" statements in the body
* [ ] `DOC601` Class docstring contains fewer class attributes than actual class attributes
* [ ] `DOC602` Class docstring contains more class attributes than in actual class attributes
* [ ] `DOC603` Class docstring attributes are different from actual class attributes
* [ ] `DOC604` Attributes are the same in docstring and class def, but are in a different order
* [ ] `DOC605` Attribute names match, but type hints in these attributes do not match

**Comment by RubenVanEldik** [comment]
I really like these new rules, they provide a lot of extra guardrails for docstrings!

When trying to start using rule `DOC202` I ran into a bug. When defining an abstract method, the rule raises an error, because nothing is returned (which makes sense, since it's an empty method, the real code is in the subclasses).

As a small example, an error is raised for this method:

```py
    @property
    @abc.abstractmethod
    def realised_prices(self) -> pd.DataFrame:
        """
        The realised prices of the market.

        Returns:
            The realised prices as a DataFrame
        """
```

**Comment by tjkuson** [comment]
Does the Astral team have an opinion on the implementation of DOC109, DOC110, and DOC111? Separating DOC109 and DOC110 seems inconsistent with how D417 works. Seeing as DOC111 is just the inverse behaviour, my intuition is that it would make the most sense for these to be consolidated into one configurable rule.

**Comment by ddelange** [comment]
That config also overlaps with the strikethrough rules covered by ANN. We currently ignore ANN because we run mypy. When the function is fully typehinted, the docstring doesn't need any types, sphinx will still render the docstring including arg and return types [for example](https://mapply.readthedocs.io/en/stable/_code_reference/mapply.html#mapply.mapply.mapply).

**Comment by Ben-Epstein** [comment]
Does anyone know if other linting tools already implemented by ruff support the `DOC6xx` set of rules? I'm a bit surprised but I couldn't find them in pydocstyle, pylint, or flake8.

**Comment by KyleKing** [comment]
I really like these rules as well!

However, it would be great if the rules only applied to public functions and methods as indicated by a `_` prefix for private.

For example, I wouldn't want to waste time adding a complete and correct docstring when the function is private and ignored by most documentation generators.

```py
def _escape(text: str) -> str:
    """Hypothetical function that does some internal business logic."""
    if "&&" in text:
        raise CustomError()
    return text.replace(";", "\;")
```

I believe a workaround today would be to move any private functions into a file prefixed by `_` and use a file pattern ignore (e.g. `_*`), but it would be nice to avoid extra refactoring and file-splitting

**Comment by augustelalande** [comment]
@KyleKing 
> However, it would be great if the rules only applied to public functions and methods as indicated by a _ prefix for private.
This makes sense to me. We should add a configuration option, to optionally allow/disallow the DOC rules on private functions.

Unfortunatelly, I won't be able to get to this any time soon. If you're up to it, feel free to open a PR.

**Comment by tjkuson** [comment]
> @KyleKing
> 
> > However, it would be great if the rules only applied to public functions and methods as indicated by a _ prefix for private.
> 
> This makes sense to me. We should add a configuration option, to optionally allow/disallow the DOC rules on private functions.
> 
> Unfortunatelly, I won't be able to get to this any time soon. If you're up to it, feel free to open a PR.

I can work on that (and ignore nested functions as well).

**Comment by ddelange** [comment]
regarding nested functions, private functions etc, here's a nice source of configs (from interrogate, a docstring coverage tool):

```ini
ignore-init-method = false
ignore-init-module = false
ignore-magic = false
ignore-semiprivate = false
ignore-private = false
ignore-property-decorators = false
ignore-module = false
ignore-nested-functions = false
ignore-nested-classes = false
ignore-setters = false
ignore-overloaded-functions = false
```

[source](https://github.com/econchick/interrogate?tab=readme-ov-file#configuration)

**Comment by ddelange** [comment]
And personally I'd love an `ignore-oneline-docstrings` option (darglint's `strictness=short`).

**Comment by tjkuson** [comment]
> And personally I'd love an `ignore-oneline-docstrings` option (darglint's `strictness=short`).

That particular configuration is likely blocked by #5860 – we'd need a member of the Astral team to decide what level of docstring configuration they'd like and how (e.g., if the configuration would apply to a subset of docstring diagnostics, or all of them).

**Comment by tjkuson** [comment]
`DOC303` is covered by `DOC202`

**Comment by Apakottur** [comment]
Pydoclint introduced [DOC503](https://github.com/jsh9/pydoclint/pull/161) for checking exceptions a few weeks ago. It's covered by DOC501/DOC502 in Ruff (which are still in preview), but perhaps worth adding to the list in this issue as completed?

**Comment by ssbarnea** [comment]
@Apakottur Yeah, and I already raised https://github.com/jsh9/pydoclint/issues/171 about it. Apparently ruff is not affected by it.

**Comment by jsh9** [comment]
> > [@KyleKing](https://github.com/KyleKing)
> > > However, it would be great if the rules only applied to public functions and methods as indicated by a _ prefix for private.
> > 
> > 
> > This makes sense to me. We should add a configuration option, to optionally allow/disallow the DOC rules on private functions.
> > Unfortunatelly, I won't be able to get to this any time soon. If you're up to it, feel free to open a PR.
> 
> I can work on that (and ignore nested functions as well).

Hi @KyleKing @augustelalande @tjkuson and @ddelange :

I'm the author of _pydoclint_.  Maybe I can share some thoughts behind why pydoclint does not support omitting private functions/methods.

Suppose a linter only checks certain functions in a repo and omits other functions, we may end up with a repo with mixed docstring styles as well as some other implications.  For example:

* Public functions have nicely written docstrings, and private functions have incorrect docstrings
* Public functions have Google style docstrings (as enforced by a linter), while private functions' docstrings are in all sorts of styles
* If we change some functions from public to private, their docstrings suddenly become unchecked
* If we change some functions from private to public, we could have a surge of violations due to months of being omitted

That's why I didn't design such an exception in _pydoclint_.  That said, there are indeed ways to "be lazy":
* Don't write any docstring, then _pydoclint_ has nothing to check.  (_pydoclint_ defers to other linters such as _pydocstyle_ to enforce the existence of docstrings.)
* Write "one line" docstrings (without Args, Returns, Yields, etc. sections)
  * This is controlled by _pydoclint_'s `--skip-checking-short-docstrings` option (which is `True` by default)

**Comment by jsh9** [comment]
> I would like to enforce a certain style docstrings. I do not think it is contained in the existing rules here or in `darglint`. Suggesting it to be `DOC901`.
> 
> In theory this case could be covered by the full set of rules if you enable all: it requires docstrings to be present, and then if we use another style these parameters will not be picked up. However, I would like to make docstrings optional, and only enforce correctness and completeness then.

Hi @sbrugman , 

I'm the author of _pydoclint_.  In the latest version of _pydoclint_ (0.6.0), I added a new violation code (`DOC003`) to perform a docstring style mismatch check: for example, if you specify `--style=google`, but your functions' docstring are written in another style, you'll see `DOC003`.

More details (including caveats) are here: https://jsh9.github.io/pydoclint/style_mismatch.html

The authors of Ruff are welcome to add this new code in Ruff's roadmap.

**Comment by tjkuson** [comment]
I'll have a PR up for DOC104 this weekend

EDIT from over a year later: I don't remember why, but I never got around to posting a PR, deepest apologies. It looks like there is a PR up now though, than you @Forsigg!

**Comment by HenriBlacksmith** [comment]
Is there any plan to stabilize the pydoclint rules which have already been implemented?

**Comment by ntBre** [comment]
We've looked into stabilizing them in the previous two minor releases (0.9 and 0.10/11),  but many, or possibly all, of them had recent PRs adjusting their behavior, or open, unresolved issues. There's also an open [PR](https://github.com/astral-sh/ruff/pull/13286) adding support for Sphinx-style docstrings, which are not supported currently.

In short, I think we're a bit hesitant to stabilize them while there's still a lot of development and discussion ongoing.

**Comment by HenriBlacksmith** [comment]
Any updates on `pydoclint` support (in 2026)?

**Comment by Forsigg** [comment]
Hello! Can I take on the DOC104 implementation? I'm almost done with the draft.

UPD: PR: https://github.com/astral-sh/ruff/pull/23047

**Comment by Forsigg** [comment]
Also, if no one is against it, I would like to work on the 300 rules (for class docstrings). They are also almost done. Sorry for my broken english

UPD: PR: https://github.com/astral-sh/ruff/pull/23081
